File: Depth-First Search/Combinations/LET_no1601_Maximum Number of Achievable Transfer Requests.py
class Solution:
    def maximumRequests(self, n: int, requests: List[List[int]]) -> int:
        self.ans = 0
        count = [0] * n
        self.dfs(requests, 0, count, 0)
        return self.ans
    
    def dfs(self, requests, index, count, num):
        if index == len(requests):
            for i in count:
                if i != 0:
                    return
            self.ans = max(self.ans, num)
            return
        
        count[requests[index][0]] -= 1
        count[requests[index][1]] += 1
        self.dfs(requests, index + 1, count, num + 1)
        count[requests[index][0]] += 1
        count[requests[index][1]] -= 1
        
        self.dfs(requests, index + 1, count, num)

File: Depth-First Search/Combinations/LET_no1601_Maximum Number of Achievable Transfer Requests.cpp
class Solution {
private:
    int max_res;
public:
    int maximumRequests(int n, vector<vector<int>>& requests) {
        max_res = 0;
        vector<int> memo(n, 0);
        dfs(requests, memo, 0, 0);
        return max_res;
    }

    void dfs(vector<vector<int>>& requests, vector<int>& memo, int idx, int count) {
        if (idx == requests.size()) {
            for (int num : memo) {
                if (num != 0)
                    return;
            }
            
            max_res = max(max_res, count);
            return;
        }

        int from = requests[idx][0], to = requests[idx][1];
        memo[from]--;
        memo[to]++;
        dfs(requests, memo, idx + 1, count + 1);
        memo[to]--;
        memo[from]++;

        dfs(requests, memo, idx + 1, count);
    }
};